dart/constraint/ConstraintSolver.cpp

```diff
@@ -542,31 +576,31 @@ void ConstraintSolver::updateConstraints()
       ++contactPairMap[std::make_pair(
           contact.collisionObject1, contact.collisionObject2)];
 
-      mContactConstraints.push_back(
-          std::make_shared<ContactConstraint>(contact, mTimeStep));
+      contacts.push_back(&contact);
     }
   }
 
   // Add the new contact constraints to dynamic constraint list
-  for (const auto& contactConstraint : mContactConstraints)
   {
-    // update the slip compliances of the contact constraints based on the
-    // number of contacts between the collision objects.
-    auto& contact = contactConstraint->getContact();
-    std::size_t numContacts = 1;
-    auto it = contactPairMap.find(
-        std::make_pair(contact.collisionObject1, contact.collisionObject2));
-    if (it != contactPairMap.end())
-      numContacts = it->second;
-
-    contactConstraint->setPrimarySlipCompliance(
-        contactConstraint->getPrimarySlipCompliance() * numContacts);
-    contactConstraint->setSecondarySlipCompliance(
-        contactConstraint->getSecondarySlipCompliance() * numContacts);
-    contactConstraint->update();
-
-    if (contactConstraint->isActive())
-      mActiveConstraints.push_back(contactConstraint);
+    std::lock_guard<std::mutex> lock(gContactSurfaceHandlersMutex);
+    for (auto* contact : contacts)
+    {
+      std::size_t numContacts = 1;
+      auto it = contactPairMap.find(
+          std::make_pair(contact->collisionObject1, contact->collisionObject2));
+      if (it != contactPairMap.end())
+        numContacts = it->second;
+
+      auto contactConstraint = gContactSurfaceHandlers[this]->createConstraint(
```

Comment:
nit: The search for the handler, i.e. `gContactSurfaceHandlers[this]` could be moved outside of the loop.

Reply:
Will address this in a followup PR.